### PR TITLE
Update graph partitioning scheme

### PR DESCRIPTION
**Pull request by seunghwak**
Partially addresses Issue #1442

Update graph partitioning scheme to better control memory footprint vs concurrency trade-offs for large-scale graph processing in large clusters. This new partitioning scheme also simplifies communication patterns among GPUs which can potentially improve scalability.

### CODE REVIEW

**Comment by seunghwak** [comment]
rerun tests

**Comment by seunghwak** [comment]
rerun tests

**Review by ChuckHastings — APPROVED** [review]

**Review by afender — APPROVED** [review]
Lots of code in there. It's great but a bit challenging to review. Moving forward it would be better to have a finer PR granularity and a high-level description of the concrete steps taken to improve a given feature. We also need to start measuring improvements when it comes to performance and scalability optimizations.

**Review by aschaffer — APPROVED** [review]

File: cpp/include/experimental/detail/graph_utils.cuh [hunk 1]
```diff
@@ -56,63 +56,36 @@ rmm::device_uvector<edge_t> compute_major_degrees(
...
-  if (partition.is_hypergraph_partitioned()) {
-    status =
-      col_comm.sync_stream(handle.get_stream());  // this is neessary as local_degrees will become
-                                                  // out-of-scope once this function returns.
-  } else {
-    status =
-      row_comm.sync_stream(handle.get_stream());  // this is neessary as local_degrees will become
-                                                  // out-of-scope once this function returns.
-  }
+  // FIXME: is this necessary?
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/experimental/detail/graph_utils.cuh:85]: Yeah... you guys are correct, this is confusing.

I added the sync and comments when I wasn't aware that rmm::device_uvector's destructor is also stream ordered; the destructor will not be called before all the previous operations on the same stream gets finished.

The destructor of local_degrees will not be called before all the operations queued to the same stream before local_degrees gets out-of-scope.

So, I added the FIXME to double-check this is indeed unnecessary (e.g. if there is another host vector gets out-of-scope, that host vector is not stream ordered, so in such cases, synchronization is necessary), and delete the synchronization.

I will double check this now, and delete confusing comments and unnecessary synchronization.

File: cpp/include/experimental/graph_view.hpp [hunk 2]
```diff
@@ -334,8 +309,6 @@ class graph_view_t<vertex_t,
                bool sorted_by_global_degree_within_vertex_partition,
                bool do_expensive_check = false);
 
-  bool is_weighted() const { return adj_matrix_partition_weights_.size() > 0; }
-
   // FIXME: this should be removed once MNMG Louvain is updated to use graph primitives
   partition_t<vertex_t> get_partition() const { return partition_; }
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/experimental/graph_view.hpp:313]: Yes, this should be removed, and I will remove this.

**Comment by seunghwak** [comment]
> Lots of code in there. It's great but a bit challenging to review. Moving forward it would be better to have a finer PR granularity and a high-level description of the concrete steps taken to improve a given feature. We also need to start measuring improvements when it comes to performance and scalability optimizations.

Yeah... sorry for creating a huge PR. This was somewhat unavoidable as this changes the fundamental underlying data structure and multiple places get affected by this change. And I should fix bugs in multiple places to make this PR work. Yeah... but next time, I will try to plan ahead to avoid creating a huge PR; I will encounter a similar challenge when I need to bring DCSR/DCSC, and I will think about bringing this in multiple steps.

**Comment by seunghwak** [comment]
I think I addressed all the comments and let me know if you want me to make additional changes. Once this PR gets merged, PR https://github.com/rapidsai/cugraph/pull/1447 will become reviewable.

**Comment by seunghwak** [comment]
rerun tests

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**